test/config/utility.h

```diff
@@ -17,19 +20,54 @@ class ConfigHelper {
   // and admin connections.
   ConfigHelper(const Network::Address::IpVersion version);
 
-  // Set the upstream ports.  The size of this vector must match the number of socket addresses
-  // across all configured clusters.
-  void setUpstreamPorts(const std::vector<uint32_t>& ports);
+  typedef std::function<void(envoy::api::v2::Bootstrap&)> ConfigModifierFunction;
+  typedef std::function<void(envoy::api::v2::filter::HttpConnectionManager&)> HttpModifierFunction;
+
+  // A string for a basic buffer filter, which can be used with addFilter()
+  static const std::string DEFAULT_BUFFER_FILTER;
+  // a string for a health check filter which can be used with addFilter()
+  static const std::string DEFAULT_HEALTH_CHECK_FILTER;
+
+  // Run the final config modifiers, and then set the upstream ports based on upstream connections.
+  // This is the last operation run on |bootstrap_| before it is handed to Envoy.
+  void finalize(const std::vector<uint32_t>& ports);
 
   // Set source_address in the bootstrap bind config.
   void setSourceAddress(const std::string& address_string);
 
+  // Overwrite the first host and route for the primary listener.
+  void setDefaultHostAndRoute(const std::string& host, const std::string& route);
+
+  // Sets byte limits on upstream and downstream connections.
+  void setBufferLimits(uint32_t upstream_buffer_limit, uint32_t downstream_buffer_limit);
+
+  // Add an additional route to the configuration.
+  void addRoute(
+      const std::string& host, const std::string& route, const std::string& cluster,
+      envoy::api::v2::VirtualHost::TlsRequirementType type = envoy::api::v2::VirtualHost::NONE);
+
+  // Add an HTTP filter prior to existing filters.
+  void addFilter(const std::string& filter_yaml);
+
+  // Allows callers to do their own modification to |bootstrap_| which will be
+  // applied just before ports are modified in finalize().
+  void addConfigModifier(ConfigModifierFunction function) { config_modifiers_.push_back(function); }
+
+  // Allows callers to easily modify the HttpConnectionManager configuration.
+  // Mofidiers will be applied just before ports are modified in finalize
+  void addConfigModifier(HttpModifierFunction function);
+
   // Return the bootstrap configuration for hand-off to Envoy.
   const envoy::api::v2::Bootstrap& bootstrap() { return bootstrap_; }
 
 private:
+  void loadHttpConnectionManager(envoy::api::v2::filter::HttpConnectionManager& hcm);
+  void storeHttpConnectionManager(const envoy::api::v2::filter::HttpConnectionManager& hcm);
```

Comment:
I suspect my comments about mutation are behind the load/store interface here. It does seem a bit more awkward than just grabbing a mutable ref. I was thinking a more pure functional approach with closures, but yeah.

Reply:
Oh, ignore this, I see, it's the`Struct` encoding.